File: trabalhoVoucher.c
#include <stdio.h>

void jogar(int pilhas[], int npilhas);
void movimento_da_pessoa(int pilhas[], int npilhas);
int escolhe_pilha(int pilhas[], int npilhas);
int escolhe_palitos(int pilhas[], int pilha_escolhida);
void movimento_da_maquina(int pilhas[], int npilhas);

int main() {
    int pilhas[100];
    int npilhas;

    printf("Quantas pilhas são? ");
    scanf("%d", &npilhas);

    for (int i = 0; i < npilhas; i++) {
        printf("Quantos palitos tem a pilha %d? ", i + 1);
        scanf("%d", &pilhas[i]);
    }

    jogar(pilhas, npilhas);

    return 0;
}

void jogar(int pilhas[], int npilhas) {
    int primeiro;

    printf("Quem começa? (Digite 1 para o jogador ou 0 para a máquina) ");
    scanf("%d", &primeiro);

    if (primeiro == 1) {
        movimento_da_pessoa(pilhas, npilhas);
    }

    while (1) {
        int soma = 0;
        for (int i = 0; i < npilhas; i++) {
            soma += pilhas[i];
        }

        if (soma == 0) {
            printf("A máquina ganhou!\n");
            break;
        }

        movimento_da_maquina(pilhas, npilhas);

        soma = 0;
        for (int i = 0; i < npilhas; i++) {
            soma += pilhas[i];
        }

        if (soma == 0) {
            printf("Você ganhou!\n");
            break;
        }

        movimento_da_pessoa(pilhas, npilhas);
    }
}

void movimento_da_pessoa(int pilhas[], int npilhas) {
    int pilhas_com_palitos[100];
    int npilhas_com_palitos = 0;

    for (int i = 0; i < npilhas; i++) {
        if (pilhas[i] > 0) {
            pilhas_com_palitos[npilhas_com_palitos] = i + 1;
            npilhas_com_palitos++;
        }
    }

    int pilha_escolhida = escolhe_pilha(pilhas_com_palitos, npilhas_com_palitos);
    int npalitos = escolhe_palitos(pilhas, pilha_escolhida);

    pilhas[pilha_escolhida - 1] -= npalitos;

    printf("Você tirou palitos. As pilhas agora são: ");
    for (int i = 0; i < npilhas; i++) {
        printf("%d ", pilhas[i]);
    }
    printf("\n");
}

int escolhe_pilha(int pilhas_com_palitos[], int npilhas_com_palitos) {
    int pilha;

    while (1) {
        printf("De qual pilha você quer tirar palitos? Escolha entre ");
        for (int i = 0; i < npilhas_com_palitos; i++) {
            printf("%d ", pilhas_com_palitos[i]);
        }
        printf(": ");
        scanf("%d", &pilha);

        int pilha_valida = 0;
        for (int i = 0; i < npilhas_com_palitos; i++) {
            if (pilha == pilhas_com_palitos[i]) {
                pilha_valida = 1;
                break;
            }
        }

        if (pilha_valida) {
            return pilha;
        } else {
            printf("Essa pilha não tem palitos\n");
        }
    }
}

int escolhe_palitos(int pilhas[], int pilha_escolhida) {
    int palitos;

    while (1) {
        printf("Quantos palitos você quer tirar? ");
        scanf("%d", &palitos);

        if (palitos > pilhas[pilha_escolhida - 1]) {
            printf("Essa pilha não tem tantos palitos\n");
        } else if (palitos < 1) {
            printf("Você deve tirar pelo menos um palito\n");
        } else {
            return palitos;
        }
    }
}

void movimento_da_maquina(int pilhas[], int npilhas) {
    int NimSum = 0;
    int pilha_com_palitos = 0;

    for (int i = 0; i < npilhas; i++) {
        NimSum ^= pilhas[i];
        if (pilhas[i] > 0) {
            pilha_com_palitos = i + 1;
        }
    }

    if (NimSum == 0) {
        pilhas[pilha_com_palitos - 1]--;
        printf("A máquina tirou palitos. As pilhas agora são: ");
        for (int i = 0; i < npilhas; i++) {
            printf("%d ", pilhas[i]);
        }
        printf("\n");
    } else {
        for (int i = 0; i < npilhas; i++) {
            if ((pilhas[i] ^ NimSum) < pilhas[i]) {
                pilhas[i] -= NimSum;
                printf("A máquina tirou palitos. As pilhas agora são: ");
                for (int j = 0; j < npilhas; j++) {
                    printf("%d ", pilhas[j]);
                }
                printf("\n");
                break;
            }
        }
    }
}

File: trabalhoVoucher.py
def main():
    pilhas = definir_jogo()
    jogar(pilhas)


def definir_jogo():
    pilhas = []
    npilhas = input("Quantas pilhas são? ")
    for i in range(0, int(npilhas)):
        npalitos = input("Quantos palitos tem a pilha " + str(i+1) + "? ")
        pilhas.append(int(npalitos))
    
    return pilhas


def jogar(pilhas):
    primeiro = int(input("Quem começa? (Digite 1 para o jogador ou 0 para a máquina)"))

    if primeiro == 1:
        movimento_da_pessoa(pilhas)
    while sum(pilhas) > 0:
        pilhas = movimento_da_maquina(pilhas)
        if sum(pilhas) == 0:
            print("A máquina ganhou!")
            break
        pilhas = movimento_da_pessoa(pilhas)
        if sum(pilhas) == 0:
            print("Você ganhou!")
            break


def movimento_da_pessoa(pilhas):
    #separe as pilhas com palitos
    pilhas_com_palitos = []
    for i in range(0, len(pilhas)):
        if pilhas[i] > 0:
            pilhas_com_palitos.append(i+1)
    #escolhe a pilha a se retirar os palitos
    pilha_escolhida = escolhe_pilha(pilhas_com_palitos)

    #escolhe a quantidade de palitos a se retirar
    npalitos = escolhe_palitos(pilhas, pilha_escolhida)
    
    
    pilhas[pilha_escolhida] = pilhas[pilha_escolhida] - npalitos
    
    print("Você tirou palitos. As pilhas agora são: " + str(pilhas))

    return pilhas

def escolhe_pilha(pilhas_com_palitos):
    pilha = int(input("De qual pilha você quer tirar palitos? Escolha entre " + str(pilhas_com_palitos)+ ": "))
    if pilha not in pilhas_com_palitos:
        print("Essa pilha não tem palitos")
        return escolhe_pilha(pilhas_com_palitos)
    else:
        return pilha - 1

def escolhe_palitos(pilhas, pilha_escolhida):
    palitos = int(input("Quantos palitos você quer tirar? "))
    if palitos > pilhas[pilha_escolhida]:
        print("Essa pilha não tem tantos palitos")
        return escolhe_palitos(pilhas, pilha_escolhida)
    elif palitos < 1:
        print("Você deve tirar pelo menos um palito")
        return escolhe_palitos(pilhas, pilha_escolhida)
    else: 
        return palitos

def movimento_da_maquina(pilhas):
    NimSum = 0
    pilha_com_palitos = 0
    for i in range(0, len(pilhas)):
        NimSum = NimSum ^ pilhas[i]
        if pilhas[i] > 0:
            pilha_com_palitos = i
    if NimSum == 0:
        pilhas[pilha_com_palitos] = pilhas[pilha_com_palitos] - 1
        print("A máquina tirou palitos. As pilhas agora são: " + str(pilhas))
    else:
        for i in range(0, len(pilhas)):
            if pilhas[i] ^ NimSum < pilhas[i]:
                pilhas[i] = pilhas[i] ^ NimSum
                print("A máquina tirou palitos. As pilhas agora são: " + str(pilhas))
                break

    return pilhas

if __name__ == "__main__":
    main()